src/rust/bin/inspect/main.rs

```diff
@@ -0,0 +1,127 @@
+        load_parameters(
+            &folder.join("sapling-spend.params"),
+            &folder.join("sapling-output.params"),
+            Some(&folder.join("sprout-groth16.params")),
+        )
+    };
+    static ref ORCHARD_VK: orchard::circuit::VerifyingKey = orchard::circuit::VerifyingKey::build();
+}
+
+#[derive(Debug, Options)]
+struct CliOptions {
+    #[options(help = "Print this help output")]
+    help: bool,
+
+    #[options(free, required, help = "String or hex-encoded bytes to inspect")]
+    data: String,
+
+    #[options(
+        free,
+        help = "JSON object with keys corresponding to requested context information"
+    )]
+    context: Option<Context>,
+}
+
+fn main() {
+    let args = env::args().collect::<Vec<_>>();
+    let mut opts =
+        CliOptions::parse_args(&args[1..], ParsingStyle::default()).unwrap_or_else(|e| {
+            eprintln!("{}: {}", args[0], e);
+            process::exit(2);
+        });
+
+    if opts.help_requested() {
+        println!("Usage: {} data [context]", args[0]);
+        println!();
+        println!("{}", CliOptions::usage());
+        return;
+    }
+
+    let lang = zip339::Language::English;
+
+    if let Ok(mnemonic) = zip339::Mnemonic::from_phrase_in(lang, &opts.data) {
+        opts.data.zeroize();
+        keys::inspect_mnemonic(mnemonic, lang, opts.context);
+    } else if let Ok(bytes) = hex::decode(&opts.data) {
+        inspect_bytes(bytes, opts.context);
+    } else if let Ok(addr) = ZcashAddress::try_from_encoded(&opts.data) {
+        address::inspect(addr);
+    } else {
+        // Unknown data format.
+        eprintln!("String does not match known Zcash data formats.");
+        process::exit(2);
+    }
+}
+
+/// Ensures that the given reader completely consumes the given bytes.
+fn complete<F, T>(bytes: &[u8], f: F) -> Option<T>
+where
+    F: FnOnce(&mut Cursor<&[u8]>) -> io::Result<T>,
+{
+    let mut cursor = Cursor::new(bytes);
+    let res = f(&mut cursor);
+    res.ok().and_then(|t| {
+        if cursor.position() >= bytes.len() as u64 {
+            Some(t)
+        } else {
+            None
+        }
+    })
+}
+
+fn inspect_bytes(bytes: Vec<u8>, context: Option<Context>) {
+    if let Some(block) = complete(&bytes, |r| block::Block::read(r)) {
+        block::inspect(&block, context);
+    } else if let Some(header) = complete(&bytes, |r| BlockHeader::read(r)) {
+        block::inspect_header(&header, context);
+    } else if let Some(tx) = complete(&bytes, |r| Transaction::read(r, BranchId::Sprout)) {
+        // TODO: Take the branch ID used above from the context if present.
+        transaction::inspect(tx, context);
+    } else {
+        // It's not a known variable-length format. check fixed-length data formats.
+        inspect_fixed_length_bytes(bytes);
+    }
+}
+
+fn inspect_fixed_length_bytes(bytes: Vec<u8>) {
+    match bytes.len() {
+        32 => {
+            eprintln!("This is most likely a hash of some sort.");
```

Comment:
```suggestion
            eprintln!("This is most likely a hash of some sort, or maybe a commitment, nullifier, or public key.");
```

Reply:
It's unlikely to be a key (for all the key types we use in Zcash, we have encodings that are not hex-encoded 32 bytes), but you're correct that it could be a commitment or nullifier (which I was bundling up as "hash of some sort" but it's probably more useful to split this out).